src/Microsoft.Extensions.OData.Migration/QueryNodeTranslator.cs

```diff
@@ -0,0 +1,364 @@
+﻿// ------------------------------------------------------------------------------
+// <copyright company="Microsoft Corporation">
+//     Copyright © Microsoft Corporation. All rights reserved.
+// </copyright>
+// ------------------------------------------------------------------------------
+
+namespace Microsoft.Extensions.OData.Migration
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using Microsoft.OData.UriParser;
+    using Microsoft.OData.Edm;
+    using System.Collections.ObjectModel;
+    using System.Globalization;
+
+    internal class QueryNodeTranslator : Data.OData.Query.SemanticAst.QueryNodeVisitor<QueryNode>
+    {
+        private readonly IEdmModel v4model;
+
+        /// <summary>
+        /// Creates a model-aware instance of the QueryNodeTranslator.
+        /// </summary>
+        /// <param name="v4model">The v4 model to use for query validation</param>
+        public QueryNodeTranslator(IEdmModel v4model)
+        {
+            this.v4model = v4model;
+        }
+
+        /// <summary>
+        /// Translates a V3 AllNode to V4 AllNode.  Visiting child nodes need to be marked dynamic
+        /// because method overloading for overridden methods works differently than usual.
+        /// </summary>
+        /// <param name="nodeIn">V3 AllNode</param>
+        /// <returns>V4 AllNode</returns>
+        public override QueryNode Visit(Data.OData.Query.SemanticAst.AllNode nodeIn)
+        {
+            IEnumerable<RangeVariable> translated = nodeIn.RangeVariables.Select(r => TranslateRangeVariable(r));
+            AllNode ret = new AllNode(new Collection<RangeVariable>(translated.ToList()), TranslateRangeVariable(nodeIn.CurrentRangeVariable));
+            ret.Source = Visit((dynamic)nodeIn.Source) as CollectionNode;

```

Comment:
If CollectionNode is expected, I wonder if direct type casting should be used so it will error out loud. Same below